src/Compilers/CSharp/Portable/SourceGeneration/SyntaxValueProviderExtensions.cs

```diff
@@ -0,0 +1,262 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Transactions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.SourceGeneration;
+
+using Aliases = ArrayBuilder<(string aliasName, string symbolName)>;
+
+internal static class SyntaxValueProviderExtensions
+{
+    /// <summary>
+    /// Returns all syntax nodes of type <typeparamref name="T"/> if that node has an attribute on it that could
+    /// possibly bind to the provided <paramref name="attributeName"/>. <paramref name="attributeName"/> should be the
+    /// simple, non-qualified, name of the attribute, including the <c>Attribute</c> suffix, and not containing any
+    /// generics, containing types, or namespaces.  For example <c>CLSCompliantAttribute</c> for <see
+    /// cref="System.CLSCompliantAttribute"/>.
+    /// <para/> This provider understands <see langword="using"/> aliases and will find matches even when the attribute
+    /// references an alias name.  For example, given:
+    /// <code>
+    /// using XAttribute = CLSCompilation;
+    /// [X]
+    /// class C { }
+    /// </code>
+    /// Then
+    /// <c>context.SyntaxProvider.CreateSyntaxProviderForAttribute&lt;ClassDeclarationSyntax&gt;("CLSCompliant")</c>
+    /// will find the <c>C</c> class.
+    /// </summary>
+    public static IncrementalValuesProvider<T> CreateSyntaxProviderForAttribute<T>(this SyntaxValueProvider provider, string attributeName)
+        where T : SyntaxNode
+    {
+        // Create a provider that provides (and updates) the global aliases for any particular file when it is edited.
+        var individualFileGlobalAliasesProvider = provider.CreateSyntaxProvider(
+            (n, _) => n is CompilationUnitSyntax,
+            (context, _) => GetGlobalAliasesInCompilationUnit((CompilationUnitSyntax)context.Node));
+
+        // Create an aggregated view of all global aliases across all files.  This should only update when an individual
+        // file changes its global aliases.
+        var allUpGlobalAliasesProvider = individualFileGlobalAliasesProvider
+            .Collect().Select((arrays, _) => GlobalAliases.Create(arrays.SelectMany(a => a.AliasAndSymbolNames).ToImmutableArray()));
+
+        // Create a syntax provider for every compilation unit.
+        var compilationUnitProvider = provider.CreateSyntaxProvider(
+            (n, _) => n is CompilationUnitSyntax,
+            (context, _) => (CompilationUnitSyntax)context.Node);
+
+        // Combine the two providers so that we reanalyze every file if the global aliases change, or we reanalyze a
+        // particular file when it's compilation unit changes.
+        var compilationUnitAndGlobalAliasesProvider = compilationUnitProvider.Combine(allUpGlobalAliasesProvider);
```

Comment:
similarly, if a change happens and the aggregrated global aliases list is unchanged, i'm hoping this provider will not rerun its transform for all compilation units.

Reply:
In this case would it change when e.g ``DesignTime`` / ``normal`` compilation unit type changes (when no code has changed between the ``DesignTime`` and the normal compilation unit?

Reply:
If no code changes, then it won't rerun.